activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,27 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def unchanged_and_covered_by_unique_index?(klass, record, attribute)
+        return false if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        return false unless attributes_scope?(record, scope)
+
+        attributes = scope + [attribute.to_s]
+        return false if attributes.any? { |attribute| record.attribute_changed?(attribute) }
+
+        klass.connection.schema_cache.indexes(klass.table_name).any? do |index|
+          index.unique &&
+            index.where.nil? &&
+            index.columns.size <= attributes.size &&
+            attributes[0, index.columns.size] == index.columns
```

Comment:
The index doesn't need to be in the same order than `attributes`. e.g.  index `[:foo, :bar, :baz]` works if `attributes == [:bar, :baz, :foo]`.

I think something like: 
```
columns = index.columns.first(attributes.size)
columns.sort == attributes.sort
``` 

It may also make sense to push part of this logic in `schema_cache`.